starknet-foundry/crates/cheatable-starknet/src/rpc.rs

```diff
@@ -0,0 +1,505 @@
+            VmExecutionContext,
+        },
+        common_hints::HintExecutionResult,
+        contract_class::{ContractClass, ContractClassV1, EntryPointV1},
+        deprecated_syscalls::DeprecatedSyscallSelector,
+        entry_point::{
+            CallEntryPoint, CallInfo, CallType, EntryPointExecutionContext,
+            EntryPointExecutionResult, ExecutionResources, FAULTY_CLASS_HASH,
+        },
+        errors::{EntryPointExecutionError, PreExecutionError, VirtualMachineExecutionError},
+        execution_utils::{felt_to_stark_felt, Args},
+        syscalls::{
+            hint_processor::{SyscallExecutionError, SyscallHintProcessor},
+            EmptyRequest, GetExecutionInfoResponse,
+        },
+    },
+    state::{cached_state::CachedState, state_api::State},
+};
+use cairo_felt_blockifier::Felt252;
+use cairo_lang_casm::{
+    hints::{Hint, StarknetHint},
+    operand::{BinOpOperand, DerefOrImmediate, Operation, Register, ResOperand},
+};
+use starknet_api::{
+    core::{ClassHash, ContractAddress, EntryPointSelector, PatriciaKey},
+    deprecated_contract_class::EntryPointType,
+    hash::{StarkFelt, StarkHash},
+    patricia_key,
+    transaction::{Calldata, TransactionVersion},
+};
+
+use blockifier::execution::syscalls::{
+    SyscallRequest, SyscallRequestWrapper, SyscallResponse, SyscallResponseWrapper, SyscallResult,
+};
+type SyscallSelector = DeprecatedSyscallSelector;
+
+pub enum CallContractOutput {
+    Success { ret_data: Vec<Felt252> },
+    Panic { panic_data: Vec<Felt252> },
+}
+
+pub struct CheatedState {
+    pub rolled_contracts: HashMap<ContractAddress, Felt252>,
+}
+
+// This can mutate state, the name of the syscall is not very good
+pub fn call_contract(
+    contract_address: &Felt252,
+    entry_point_selector: &Felt252,
+    calldata: &[Felt252],
+    blockifier_state: &mut CachedState<DictStateReader>,
+    cheated_state: &CheatedState,
+) -> Result<CallContractOutput> {
+    let contract_address = ContractAddress(PatriciaKey::try_from(StarkFelt::new(
+        contract_address.to_be_bytes(),
+    )?)?);
+    let entry_point_selector =
+        EntryPointSelector(StarkHash::new(entry_point_selector.to_be_bytes())?);
+    let account_address = ContractAddress(patricia_key!(TEST_ACCOUNT_CONTRACT_ADDRESS));
+    let calldata = Calldata(Arc::new(
+        calldata
+            .iter()
+            .map(|data| StarkFelt::new(data.to_be_bytes()))
+            .collect::<Result<Vec<_>, _>>()?,
+    ));
+    let mut entry_point = CallEntryPoint {
+        class_hash: None,
+        code_address: Some(contract_address),
+        entry_point_type: EntryPointType::External,
+        entry_point_selector,
+        calldata,
+        storage_address: contract_address,
+        caller_address: account_address,
+        call_type: CallType::Call,
+        initial_gas: u64::MAX,
+    };
+
+    let mut resources = ExecutionResources::default();
+    let account_context = build_transaction_context();
```

Comment:
`build_transaction_context`
always sets nonce to `0`. It is a inconsistent behaviour, can we create an issue to fix it?

Reply:
https://github.com/foundry-rs/starknet-foundry/issues/258